tools-and-tests/tools/src/main/java/org/hiero/block/tools/capacity/NetworkCapacityClient.java

```diff
@@ -0,0 +1,241 @@
+                .filter(path -> path.toString().endsWith(".blk.gz"))
+                .sorted()
+                .collect(Collectors.toList());
+
+        if (blockFiles.isEmpty()) {
+            System.err.printf("No block files found in %s%n", recordingFolder);
+            return;
+        }
+
+        System.out.printf("Found %d block files to stream.%n", blockFiles.size());
+
+        CompletableFuture<Void> streamingComplete = new CompletableFuture<>();
+        ResponseHandler responseHandler = new ResponseHandler(streamingComplete);
+        Pipeline<PublishStreamRequest> requestPipeline = publishClient.publishBlockStream(responseHandler);
+
+        for (Path blockFile : blockFiles) {
+            try {
+                streamBlockFile(blockFile, requestPipeline);
+                // Optional pacing; consider removing for tight windows
+                LockSupport.parkNanos(10_000_000);
+            } catch (RuntimeException | ParseException e) {
+                System.err.printf("Error streaming file %s due to %s%n", blockFile, e);
+            }
+        }
+
+        requestPipeline.onComplete();
+
+        try {
+            streamingComplete.get(Duration.ofMinutes(5).toMillis(), TimeUnit.MILLISECONDS);
+        } catch (RuntimeException | InterruptedException | ExecutionException | TimeoutException e) {
+            System.err.printf("Error waiting for stream completion due to %s%n", e);
+        }
+    }
+
+    private void streamBlockFile(Path blockFile, Pipeline<PublishStreamRequest> requestPipeline)
+            throws IOException, ParseException {
+        byte[] blockData;
+        try (final var gzipInputStream = new GZIPInputStream(Files.newInputStream(blockFile))) {
+            blockData = gzipInputStream.readAllBytes();
+        }
+        Block block = Block.PROTOBUF.parse(Bytes.wrap(blockData));
+        long blockNumber = block.items().getFirst().blockHeader().number();
+
+        List<BlockItem> items = block.items();
+        System.out.printf(
+                "Streaming block from file: %s (%d items)%n", blockFile.getFileName(), items.size());
+
+        final long maxMessageSizeBytes = config.serverMaxMessageSizeBytes();
+        int i = 0;
+        int batchCount = 0;
+
+        while (i < items.size()) {
+            List<BlockItem> batch = new ArrayList<>();
+            long currentBatchSizeBytes = 0;
+
+            // Build batch until we approach the max message size
+            while (i < items.size()) {
+                BlockItem item = items.get(i);
+                long itemSizeBytes = BlockItem.PROTOBUF.toBytes(item).length();
+
+                // If adding this item would exceed max size and batch is not empty, break
+                if (currentBatchSizeBytes + itemSizeBytes > maxMessageSizeBytes && !batch.isEmpty()) {
+                    break;
+                }
+
+                batch.add(item);
+                currentBatchSizeBytes += itemSizeBytes;
+                i++;
+            }
```

Comment:
Looks like this is "build a batch" but it's a little wonky due to needing to measure the item size.
The following is a possible alternative that takes advantage of some performance optimization in the PBJ codecs.
```suggestion
            while (batch.isEmpty() || currentBatchSizeBytes + BlockItem.PROTOBUF.measureRecord(items.get(i) <= maxMessageSizeBytes) {
                BlockItem item = items.get(i++);
                batch.add(item);
                currentBatchSizeBytes += BlockItem.PROTOBUF.measureRecord(item);
            }
```